In [1]:
## import the dependencies
## dependencies are important because instead of writing a bunch of code to do a certain task from scratch, someone
###else or another company already created it for you to use

import json
import pandas as pd
import numpy as np

In [2]:
## create the file path to your json file
## its currently saved in the module 9 folder inside surfs_up

file_dir = 'wikipedia-movies.json'

In [3]:
## we don't want to load messy json data into a df, so we need to create a list of dictionaries

with open(file_dir, mode='r') as file:
    wiki_movies_raw = json.load(file)

In [4]:
len(wiki_movies_raw)

7311

In [5]:
## see the first 5 records
## square bracket is a list

wiki_movies_raw[:5]

[{'url': 'https://en.wikipedia.org/wiki/The_Adventures_of_Ford_Fairlane',
  'year': 1990,
  'imdb_link': 'https://www.imdb.com/title/tt0098987/',
  'title': 'The Adventures of Ford Fairlane',
  'Directed by': 'Renny Harlin',
  'Produced by': ['Steve Perry', 'Joel Silver'],
  'Screenplay by': ['David Arnott', 'James Cappe', 'Daniel Waters'],
  'Story by': ['David Arnott', 'James Cappe'],
  'Based on': ['Characters', 'by Rex Weiner'],
  'Starring': ['Andrew Dice Clay',
   'Wayne Newton',
   'Priscilla Presley',
   'Lauren Holly',
   'Morris Day',
   'Robert Englund',
   "Ed O'Neill"],
  'Narrated by': 'Andrew "Dice" Clay',
  'Music by': ['Cliff Eidelman', 'Yello'],
  'Cinematography': 'Oliver Wood',
  'Edited by': 'Michael Tronick',
  'Productioncompany ': 'Silver Pictures',
  'Distributed by': '20th Century Fox',
  'Release date': ['July 11, 1990', '(', '1990-07-11', ')'],
  'Running time': '102 minutes',
  'Country': 'United States',
  'Language': 'English',
  'Budget': '$20 million',


In [6]:
## see the last 5 records by using negative index slices
wiki_movies_raw[-5]

{'url': 'https://en.wikipedia.org/wiki/Holmes_%26_Watson',
 'year': 2018,
 'imdb_link': 'https://www.imdb.com/title/tt1255919/',
 'title': 'Holmes & Watson',
 'Directed by': 'Etan Cohen',
 'Produced by': ['Will Ferrell',
  'Adam McKay',
  'Jimmy Miller',
  'Clayton Townsend'],
 'Screenplay by': 'Etan Cohen',
 'Based on': ['Sherlock Holmes',
  'and',
  'Dr. Watson',
  'by',
  'Sir Arthur Conan Doyle'],
 'Starring': ['Will Ferrell',
  'John C. Reilly',
  'Rebecca Hall',
  'Rob Brydon',
  'Steve Coogan',
  'Ralph Fiennes'],
 'Music by': 'Mark Mothersbaugh',
 'Cinematography': 'Oliver Wood',
 'Edited by': 'Dean Zimmerman',
 'Productioncompanies ': ['Columbia Pictures',
  'Gary Sanchez Productions',
  'Mosaic Media Group',
  'Mimran Schur Pictures'],
 'Distributed by': 'Sony Pictures Releasing',
 'Release date': ['December 25, 2018',
  '(',
  '2018-12-25',
  ')',
  '(United States)'],
 'Running time': '90 minutes',
 'Country': 'United States',
 'Language': 'English',
 'Budget': '$42 million

In [7]:
## checking random records in the middle
wiki_movies_raw[3600:3605]

[{'url': 'https://en.wikipedia.org/wiki/Benji:_Off_the_Leash!',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0315273/',
  'title': 'Benji: Off the Leash!',
  'Directed by': 'Joe Camp',
  'Written by': 'Joe Camp',
  'Starring': ['Benji', 'Nick Whitaker', 'Shaggy', 'Gypsy the Cockatoo'],
  'Music by': 'Antonio di Lorenzo',
  'Productioncompany ': 'Mulberry Square Productions',
  'Distributed by': 'Mulberry Square Productions',
  'Release date': ['March 26, 2004', '(', '2004-03-26', ')'],
  'Running time': '97 min',
  'Country': 'United States',
  'Language': 'English',
  'Box office': '$3,817,362'},
 {'url': 'https://en.wikipedia.org/wiki/The_Best_Thief_in_the_World',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0389796/',
  'title': 'The Best Thief in the World',
  'Directed by': 'Jacob Kornbluth',
  'Produced by': ['Tim Perrell', 'Nicola Usborne'],
  'Written by': 'Jacob Kornbluth',
  'Starring': ['Marc Rozendaal',
   'Michael Silverman',
   'David Warsh

In [8]:
## kaggle data can be imported directly to df because they are flat-files (csv)
kaggle_metadata = pd.read_csv('movies_metadata.csv', low_memory=False)
ratings = pd.read_csv('ratings.csv')

In [9]:
## inspecting the data
kaggle_metadata.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [10]:
ratings.head()

userId  movieId  rating   timestamp
0       1      110     1.0  1425941529
1       1      147     4.5  1425942435
2       1      858     5.0  1425941523
3       1     1221     5.0  1425941546
4       1     1246     5.0  1425941556

### initial investigation
* easiest way to find errors, just pretend there aren't any and import data into a dataframe

In [11]:
## convert the wiki_movies_raw data to a dataframe
wiki_movies_df = pd.DataFrame(wiki_movies_raw)

In [12]:
wiki_movies_df.head()

url    year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990.0   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990.0   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990.0   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990.0   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990.0   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Predecessor  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...         NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...         NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...         NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...         NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...         NaN   

  Founders Area served Products Services Russian Hebrew Revenue  \
0      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
1      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
2      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
3      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
4      NaN         NaN      NaN      NaN     NaN    NaN     NaN   

  Operating income Polish  
0              NaN    NaN  
1              NaN    NaN  
2              NaN    NaN  
3              NaN    NaN  
4              NaN    NaN  

[5 rows x 193 columns]

### use the column count 
* using the row and column count, you can see there are 5 rows (expected because we're using head) but there are 193 columns

In [13]:
## convert wiki_movies_df columns to a list to see the names of all the columns
wiki_movies_df.columns.tolist()

['url',
 'year',
 'imdb_link',
 'title',
 'Directed by',
 'Produced by',
 'Screenplay by',
 'Story by',
 'Based on',
 'Starring',
 'Narrated by',
 'Music by',
 'Cinematography',
 'Edited by',
 'Productioncompany ',
 'Distributed by',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Written by',
 'Genre',
 'Theme music composer',
 'Country of origin',
 'Original language(s)',
 'Producer(s)',
 'Editor(s)',
 'Production company(s)',
 'Original network',
 'Original release',
 'Productioncompanies ',
 'Executive producer(s)',
 'Production location(s)',
 'Distributor',
 'Picture format',
 'Audio format',
 'Voices of',
 'Followed by',
 'Composer(s)',
 'Created by',
 'Also known as',
 'Opening theme',
 'No. of episodes',
 'Preceded by',
 'Author',
 'Publisher',
 'Publication date',
 'Media type',
 'Pages',
 'ISBN',
 'OCLC',
 'LC Class',
 'Cover artist',
 'Series',
 'Set in',
 'Adaptation by',
 'Suggested by',
 'Biographical data',
 'Born',
 'Died',
 'Resti

In [14]:
## need to modify the json data by redistricting it to only those entries that have a director and an imdb link. 
## can be done with list comprehension -- use to filter data

wiki_movies = [movie for movie in wiki_movies_raw
               if ('Director' in movie or 'Directed by' in movie)
                   and 'imdb_link' in movie]
len(wiki_movies)

7080

In [15]:
## make a dataframe from wiki_movies
wiki_movies_df = pd.DataFrame(wiki_movies)

In [16]:
wiki_movies_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Hepburn Literally  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...     NaN       NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...     NaN       NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...     NaN       NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...     NaN       NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...     NaN       NaN   

  Cantonese Chinese Yiddish Arabic Romanized Russian Hebrew Polish  
0       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
1       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
2       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
3       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
4       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  

[5 rows x 78 columns]

In [17]:
## look through the edited column list again
wiki_movies_df.columns.tolist()

['url',
 'year',
 'imdb_link',
 'title',
 'Directed by',
 'Produced by',
 'Screenplay by',
 'Story by',
 'Based on',
 'Starring',
 'Narrated by',
 'Music by',
 'Cinematography',
 'Edited by',
 'Productioncompany ',
 'Distributed by',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Written by',
 'Genre',
 'Theme music composer',
 'Country of origin',
 'Original language(s)',
 'Producer(s)',
 'Editor(s)',
 'Production company(s)',
 'Original network',
 'Original release',
 'Productioncompanies ',
 'Executive producer(s)',
 'Production location(s)',
 'Distributor',
 'Picture format',
 'Audio format',
 'Voices of',
 'Followed by',
 'Composer(s)',
 'Created by',
 'Also known as',
 'Opening theme',
 'No. of episodes',
 'Preceded by',
 'Adaptation by',
 'Suggested by',
 'Traditional',
 'Mandarin',
 'Released',
 'Recorded',
 'Venue',
 'Length',
 'Label',
 'Director',
 'Producer',
 'Animation by',
 'Color process',
 'Simplified',
 'French',
 'Camera setup'

In [18]:
## looking through the list of columns, there are a lot of languages and there is also one called
## no. of episodes -- this means there are some that are not movies and may just be a television show
## create a list comprehension to FILTER the data

wiki_movies = [movie for movie in wiki_movies_raw
               if ('Director' in movie or 'Directed by' in movie)
                  and 'imdb_link' in movie
                  and 'No. of episodes' not in movie]
wiki_movies_df = pd.DataFrame(wiki_movies)

In [19]:
wiki_movies_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Hepburn Literally  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...     NaN       NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...     NaN       NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...     NaN       NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...     NaN       NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...     NaN       NaN   

  Cantonese Chinese Yiddish Arabic Romanized Russian Hebrew Polish  
0       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
1       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
2       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
3       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
4       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  

[5 rows x 75 columns]

### practicing scope of variables
* we can create a local variable (created inside the function) with the same name as a global variable (created outside function)

In [20]:
x = 'global value' ### this variable was created outside the function

def foo(): ### def means you're defining the function
    x = 'local value' ### this is a local variable but you still named it x
    print(x) ### print the variable
    
foo() ### calling the defined function (i think)
print(x) ### print the global value bc its outside

local value
global value


In [21]:
my_list = [1,2,3] ## creating the global variable bc it was created outside the function

def append_four(x): ## defining the function -- add 4 to the end of the list
    x.append(4) 
append_four(my_list)
print(my_list) 

[1, 2, 3, 4]


In [22]:
## this changed the 'my_list' variable to include the 4
## if we want to leave the original object (my_list) unchanged, then we need to make a copy

### new_list = list(old_list)
### new_dict = dict(old_dict)

### lambda function example

In [23]:
## how to square a value
square = lambda x: x * x ## (x) is the arguement (x * x) is the expression 'square' is the name
square(5) ### we're calling our named function and assigning 5 as the x in our formula

25

In [24]:
 ## we start the function by using def
def clean_movie(movie):
    movie = dict(movie) ## create a non-destructive copy. we're assigning movie to a local (inside) variable. 
    return movie

In [25]:
## call all the movies that have something written in the arabic column and is not null
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]

url  year  \
6834  https://en.wikipedia.org/wiki/The_Insult_(film)  2018   
7058   https://en.wikipedia.org/wiki/Capernaum_(film)  2018   

                                  imdb_link       title    Directed by  \
6834  https://www.imdb.com/title/tt7048622/  The Insult   Ziad Doueiri   
7058  https://www.imdb.com/title/tt8267604/   Capernaum  Nadine Labaki   

                                            Produced by  \
6834  [Rachid Bouchareb, Jean Bréhat, Julie Gayet, A...   
7058                    [Michel Merkt, Khaled Mouzanar]   

                                          Screenplay by  \
6834                                                NaN   
7058  [Nadine Labaki, Jihad Hojaily, Michelle Keserw...   

                                               Story by Based on  \
6834                                                NaN      NaN   
7058  [Georges Khabbaz, Nadine Labaki, Michelle Kese...      NaN   

                                               Starring  ... Hepburn  \
6834                       [Adel Karam, Kamel El Basha]  ...     NaN   
7058  [Zain Al Rafeea, Yordanos Shiferaw, Boluwatife...  ...     NaN   

        Literally Cantonese Chinese Yiddish       Arabic        Romanized  \
6834  Case No. 23       NaN     NaN     NaN  قضية رقم ٢٣  Qadiyya raqm 23   
7058          NaN       NaN     NaN     NaN    کفرناحوم‎              NaN   

     Russian Hebrew Polish  
6834     NaN    NaN    NaN  
7058     NaN    NaN    NaN  

[2 rows x 75 columns]

In [26]:
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]['url'] ## pull the url

6834    https://en.wikipedia.org/wiki/The_Insult_(film)
7058     https://en.wikipedia.org/wiki/Capernaum_(film)
Name: url, dtype: object

In [27]:
## sort in alphabetical order

sorted(wiki_movies_df.columns.tolist())

['Adaptation by',
 'Also known as',
 'Animation by',
 'Arabic',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cantonese',
 'Chinese',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Country of origin',
 'Created by',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Edited by',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'French',
 'Genre',
 'Hangul',
 'Hebrew',
 'Hepburn',
 'Japanese',
 'Label',
 'Language',
 'Length',
 'Literally',
 'Mandarin',
 'McCune–Reischauer',
 'Music by',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Original release',
 'Original title',
 'Picture format',
 'Polish',
 'Preceded by',
 'Produced by',
 'Producer',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Productioncompanies ',
 'Productioncompany ',
 'Recorded',
 'Release date',
 'Released',
 'Revised Romanization',
 'Romanized',
 'Running time',
 'Russian',
 'Screen story by',
 'Screenplay by',
 'Simplifie

In [28]:
wiki_movies_df['Polish'].value_counts() ### the language columns hold the alternate titles for the movies

Zimna wojna    1
Name: Polish, dtype: int64

### handling the alternative titles
Logic
* make an empty dict to hold all the alternative keys
* loop through a list of all alternative title keys
- check if the current key exists in the movie object
- if so, remove the key-value pair and add to the alternative titles dict
* after looping through every key, add the alternative titles dict to the movie object

In [29]:
## step 1: make an empty dict to hold all of the alternative titles
def clean_movie(movie):
    movie = dict(movie) # this is to create a non-destructive copy
    alt_titles = {} # empty dict -- curly brackets, empty list -- square brackets
    for key in ['Also known as','Arabic','Cantonese','Chinese','French', ## step2: loop through a list of all alt title keys
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']: ## creating another loop
        if key in movie: ### if so, remove the key-value pair and add to the alt titles dict
            alt_titles[key] = movie[key]
            movie.pop(key)
    if len(alt_titles) > 0: ## after looping through every key, add the alt titles to movie object
        movie['alt_titles'] = alt_titles
    return movie

In [30]:
## create a list of cleaned movies with a list comprehension
clean_movies = [clean_movie(movie) for movie in wiki_movies]

In [31]:
## set wiki_movies_df to be the df created from clean_movies
wiki_movies_df = pd.DataFrame(clean_movies)
sorted(wiki_movies_df.columns.tolist())

['Adaptation by',
 'Animation by',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Country of origin',
 'Created by',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Edited by',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'Genre',
 'Label',
 'Language',
 'Length',
 'Music by',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Original release',
 'Picture format',
 'Preceded by',
 'Produced by',
 'Producer',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Productioncompanies ',
 'Productioncompany ',
 'Recorded',
 'Release date',
 'Released',
 'Running time',
 'Screen story by',
 'Screenplay by',
 'Starring',
 'Story by',
 'Suggested by',
 'Theme music composer',
 'Venue',
 'Voices of',
 'Written by',
 'alt_titles',
 'imdb_link',
 'title',
 'url',
 'year']